### PR TITLE
Add snapshot reporter lambda

### PR DESCRIPTION
**Pull request by kenoir**
For https://github.com/wellcomecollection/platform/issues/4817

The output in Slack should appear as follows (top successful message if snapshot found, bottom failure message if not found).

Screenshot 2020-10-07 at 10 58 25

Note: The terraform in this PR is applied.

### CODE REVIEW

File: snapshots/snapshot_reporter/src/snapshot_reporter.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+def get_snapshots(es_client, days_to_fetch):
+    index = "snapshots"
+
+    date_filter = [
+        {"range": {"snapshotJob.requestedAt": {"gte": f"now-{days_to_fetch}d/d"}}}
+    ]
+    es_query = {"bool": {"filter": date_filter}}
+    sort_by = [{"snapshotJob.requestedAt": {"order": "desc"}}]
+
+    response = es_client.search(index=index, body={"query": es_query, "sort": sort_by})
```
Comment (alexwlchan) [inline snapshots/snapshot_reporter/src/snapshot_reporter.py:51]: It's a small thing, but I generally prefer to construct the Elasticsearch requests in one big dict, i.e.

```python
request_body = {
    "query": {
        "bool": {
            "filter": {"range": {"snapshotJob.requestedAt": {"gte": f"now-{days_to_fetch}d/d"}}}
        }
    },
    "sort": [{"snapshotJob.requestedAt": {"order": "desc"}}],
}
```

cos then (trailing commas aside) you can drop that straight into the Elastic dev tools for experimenting, and it's easier to follow through to the docs.

**Review by alexwlchan — COMMENTED** [review]
Looking good. ✨

File: snapshots/snapshot_reporter/src/snapshot_reporter.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+
+def prepare_slack_payload(snapshots, days_to_fetch):
+    def _snapshot_message(snapshot):
+        index_name = snapshot["snapshotResult"]["indexName"]
+        document_count = snapshot["snapshotResult"]["documentCount"]
+        s3_size = snapshot["snapshotResult"]["s3Size"]["bytes"]
+
+        requested_at = parser.parse(snapshot["snapshotJob"]["requestedAt"]).strftime(
+            "%A, %B %e, %I:%M %p"
```
Comment (alexwlchan) [inline snapshots/snapshot_reporter/src/snapshot_reporter.py:63]: nit: if you look at the rendered output, you'll see there's an extra space `Tuesday, October  6`. You don't want the space, you just want the leading zero trimmed. I think this does that.

```suggestion
            "%A, %B %-d, %I:%M %p"
```

File: snapshots/snapshot_reporter/src/snapshot_reporter.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        )
+
+        started_at = parser.parse(snapshot["snapshotResult"]["startedAt"])
+
+        finished_at = parser.parse(snapshot["snapshotResult"]["finishedAt"])
+
+        return "\n".join(
+            [
+                f'*"{index_name}" on {requested_at}*',
+                f"{humanize.naturalsize(s3_size)} containing {humanize.intword(document_count)} documents ({document_count} exactly)",
```
Comment (alexwlchan) [inline snapshots/snapshot_reporter/src/snapshot_reporter.py:73]: nit: consider adding commas for readability.

```suggestion
                f"{humanize.naturalsize(s3_size)} containing {humanize.intword(document_count)} documents ({humanize.intcomma(document_count)} exactly)",
```

File: snapshots/snapshot_reporter/src/snapshot_reporter.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+                f"Took {humanize.precisedelta(finished_at - started_at)}",
```
Comment (alexwlchan) [inline snapshots/snapshot_reporter/src/snapshot_reporter.py:74]: nit: is precisedelta really what we want? Telling me it took about 18 minutes is useful. Telling me it took 18 minutes and 8.83 seconds is probably too much precision.

File: snapshots/snapshot_reporter/src/snapshot_reporter.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+            ]
+        )
+
+    heading = f"Snapshots in the last {days_to_fetch} days"
```
Comment (alexwlchan) [inline snapshots/snapshot_reporter/src/snapshot_reporter.py:78]: nit: "1 days"

😢

**Review by alexwlchan — COMMENTED** [review]
Now I've seen the output, a couple of tiny suggestions. I am very picky, so do ignore if you don't think it's worth the time.

At a higher level: in my mind, this Lambda is about alerting us to issues with the snapshot reporter. So I'd change the output somewhat:

* Unlike the storage service, we're only interested in the last snapshot. The data is there if somebody wants to see how many other snapshots were taken, but something like:

    > The last snapshot was of index *works-20200923* at *5pm yesterday*. It is 457.8MB and contains 1.2 million documents. [link to reporting cluster]

* The exact number isn't important in this message, nor is the time taken.

* What if there aren't any snapshots in the reporting period? That seems something important it  should flag specially, optionally with a pre-built link to the appropriate log group in the logging cluster – see the ingest inspector for how we do this elsewhere. https://github.com/wellcomecollection/ingest-inspector/blob/master/server.py#L92-L140

File: snapshots/snapshot_reporter/src/snapshot_reporter.py (hunk 3 above)
Comment (jamesgorrie) [inline snapshots/snapshot_reporter/src/snapshot_reporter.py:73]: Why do we have the rounded and exactly amounts? Could we just have the exactly amounts?

File: snapshots/snapshot_reporter/src/snapshot_reporter.py (hunk 1 above)
Comment (kenoir) [reply snapshots/snapshot_reporter/src/snapshot_reporter.py:51]: I find parsing the whole block confusing TBH, but I see the benefit in being able to drop it into the ES console.

**Comment by kenoir** [comment]
> The exact number isn't important in this message, nor is the time taken.

I disagree, the exact number is useful when the granularity is to the nearest hundred thousand and i'm looking at the reports over a few days.

The time taken is also interesting. A very short time would be a warning something is wrong, and the time it takes is useful to know if we start to push up against visibility timeouts.

File: snapshots/snapshot_reporter/src/snapshot_reporter.py (hunk 3 above)
Comment (kenoir) [reply snapshots/snapshot_reporter/src/snapshot_reporter.py]: A number in millions without some extra formatting is hard to parse visually (e.g. how much is 11212422?). I think simply using comma notation might fix that though.

File: snapshots/snapshot_reporter/src/snapshot_reporter.py (hunk 5 above)
Comment (kenoir) [reply snapshots/snapshot_reporter/src/snapshot_reporter.py:78]: If we refer to only a single snapshot in the report (as you suggest) this goes away nicely.

File: snapshots/snapshot_reporter/src/snapshot_reporter.py (hunk 4 above)
Comment (kenoir) [reply snapshots/snapshot_reporter/src/snapshot_reporter.py:74]: I am inclined to agree that minute resolution is all that is required here.

**Comment by kenoir** [comment]
> What if there aren't any snapshots in the reporting period? That seems something important it should flag specially.

Yes, that's incoming.

File: snapshots/snapshot_reporter/src/snapshot_reporter.py (hunk 3 above)
Comment (jamesgorrie) [reply snapshots/snapshot_reporter/src/snapshot_reporter.py]: Yea, big numbers are hard to process... good point.

File: snapshots/snapshot_reporter/src/snapshot_reporter.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    if snapshots:
+        snapshot = snapshots[0]
+
+        header_block = _create_header_block(":white_check_mark: Catalogue Snapshot")
+        section_block = _create_section_block(_snapshot_message(snapshot))
+    else:
+        header_block = _create_header_block(
+            ":interrobang: Catalogue Snapshot not found"
+        )
+        section_block = _create_section_block("No snapshot found within the last day.")
```
Comment (alexwlchan) [inline snapshots/snapshot_reporter/src/snapshot_reporter.py:97]: I would include a link to the snapshot generator logs here so it's really easy to go debug, but it's not a big deal.

**Review by alexwlchan — APPROVED** [review]
Nice stuff. One suggestion, but it's not blocking.
